akka-http-core/src/main/scala/akka/http/scaladsl/ConnectionContext.scala

```diff
@@ -40,20 +90,38 @@ object ConnectionContext {
     enabledProtocols:    Option[immutable.Seq[String]],
     clientAuth:          Option[TLSClientAuth],
     sslParameters:       Option[SSLParameters]) =
-    new HttpsConnectionContext(sslContext, sslConfig = None, enabledCipherSuites, enabledProtocols, clientAuth, sslParameters)
+    new HttpsConnectionContext(Left(sslContext), sslConfig = None, enabledCipherSuites, enabledProtocols, clientAuth, sslParameters)
 
   def noEncryption() = HttpConnectionContext
 }
 
+private[http] case class Engine(create: () => SSLEngine, validate: SSLSession => Try[Unit] = _ => Success(()))
+
+/**
+ *  Context with all information needed to set up a HTTPS connection
+ *
+ * This constructor is INTERNAL API, use ConnectionContext.https instead
+ */
+@InternalApi
 final class HttpsConnectionContext(
-  val sslContext:          SSLContext,
-  val sslConfig:           Option[AkkaSSLConfig]         = None,
-  val enabledCipherSuites: Option[immutable.Seq[String]] = None,
-  val enabledProtocols:    Option[immutable.Seq[String]] = None,
-  val clientAuth:          Option[TLSClientAuth]         = None,
-  val sslParameters:       Option[SSLParameters]         = None)
+  val sslContextData:      Either[SSLContext, Option[(String, Int)] => Engine],
```

Comment:
Do we need that distinction here? Couldn't we change it to just use the engine factory and provide that directly from the context in the constructor?

Reply:
I see, it might be for backwards compat?

Reply:
Yes - all the non-deprecated methods provide `Option[(String, Int)] => Engine`, but the `SSLContext` is here to keep the old behavior for when the deprecated API's have been called.

Reply:
Revisting that part: is the only reason we need to keep the SSLContext to be able to call 

```
TLS(sslContext, connectionContext.sslConfig, ...
```

later on?

This TLS constructor also eventually calls the other one, so in theory, we could copy over the sslconfig related bits from there, right? That would allow us to decouple our deprecation from the one in Akka. On the other hand, 10.3.0 might be before Akka 2.7.0, so maybe it's not worth it.

Reply:
Could it make sense otherwise, to include the `sslConfig` param in the left side of the either as well?

Reply:
I agree this would be neater, but I think this is not possible because there would be one less default `Option` in the constructor, and while for normal methods you could add the previously-automatically-generated method for the missing generic, because this is a constructor the name is `<init>$default$6` which AFAICS you can't recreate...